### PR TITLE
[RFC] Return empty string if no working directory.

### PR DESCRIPTION
**Pull request by HiPhish**
This restores behaviour identical to Vim. If the user calls the VimScript function 'getcwd()' and the working directory cannot be found (for example because the directory has been deleted since the last time it was used) an empty string needs to be returned instead of throwing an error.

I still need to get the test working properly before this PR can be considered final.

### CODE REVIEW

File: test/functional/ex_cmds/cd_spec.lua [hunk 1]
```diff
@@ -269,3 +269,24 @@ for _, cmd in ipairs {'getcwd', 'haslocaldir'} do
...
+    lfs.mkdir(temp_dir)
+  end)
+
+  after_each(function()
+    lfs.rmdir(temp_dir)
+  end)
+
+  it('returns an empty string', function()
+    execute('cd ' .. temp_dir)
+    -- lfs.rmdir(temp_dir)
```
Comment (fwalch) [inline test/functional/ex_cmds/cd_spec.lua]: Adding a `helpers.wait()` before the `rmdir` works for me.

File: test/functional/ex_cmds/cd_spec.lua [hunk 2]
```diff
@@ -269,3 +269,24 @@ for _, cmd in ipairs {'getcwd', 'haslocaldir'} do
...
+    local working_dir = call('getcwd')
```
Comment (fwalch) [inline test/functional/ex_cmds/cd_spec.lua]: Can change to `local working_dir = cwd()`.

File: test/functional/ex_cmds/cd_spec.lua (hunk 1 above)
Comment (HiPhish) [inline test/functional/ex_cmds/cd_spec.lua]: Thank's that got it working.

**Comment by HiPhish** [comment]
OK, this should be good to go now.

File: test/functional/ex_cmds/cd_spec.lua [hunk 3]
```diff
@@ -269,3 +269,25 @@ for _, cmd in ipairs {'getcwd', 'haslocaldir'} do
...
+  after_each(function()
+    lfs.rmdir(temp_dir)
+  end)
+
+  it('returns an empty string', function()
+    execute('cd ' .. temp_dir)
+    helpers.wait()
+    lfs.rmdir(temp_dir)
+    local working_dir = cwd()
+    eq('', working_dir)
```
Comment (justinmk) [inline test/functional/ex_cmds/cd_spec.lua]: If I revert the change to eval.c, this doesn't fail correctly. It seems that `call()` ignores VimL errors. So I changed this to `helpers.eval("getcwd()")`.

Comment (ZyX-I) [inline test/functional/ex_cmds/cd_spec.lua]: I usually write such things as something like `eq(0, exc_exec('let wd = getcwd()'))`: this makes it obvious that I check for absence of exceptions and not rely on implementation details of various helper functions.

Comment (justinmk) [inline test/functional/ex_cmds/cd_spec.lua]: @ZyX-I I thought about that but I think our test functions should always fail on VimL errors. Does it seem correct to you that `call()` would be silent?

**Comment by justinmk** [comment]
Merged

**Comment by blueyed** [comment]
Thanks!

File: test/functional/ex_cmds/cd_spec.lua (hunk 3 above)
Comment (ZyX-I) [inline test/functional/ex_cmds/cd_spec.lua]: Such failure is an error (BTW, why screen test failures are errors and not failures?), `eq(0,…)` failure is a failure which is more in line with the purpose of the test. I do not see why original variant does not fail though, `nvim_call_function` should generate error in this case due to the used `try_*` pair.

Comment (justinmk) [inline test/functional/ex_cmds/cd_spec.lua]: > BTW, why screen test failures are errors and not failures?)

I don't think there's a reason. It's always bugged me.
